### PR TITLE
[feat] Telemetry(Bonk-terminal): Add telementry fee/revenue adapter

### PR DESCRIPTION
**Pull request by Shaileshkhote**
## Summary
- Adds a new fees adapter for Telemetry, the Solana web/mobile trading terminal by BONKbot.
- Tracks Telemetry fees, revenue, user fees, protocol revenue, and referral/cashback rewards on Solana.
- Website: https://telemetry.io/
- Twitter/X: https://x.com/bonkbot_io

##### Name (to be shown on DefiLlama):
Telemetry

##### Twitter Link:
https://x.com/bonkbot_io

##### Website Link:
https://telemetry.io/

##### Chain:
Solana

##### Short Description (to be shown on DefiLlama):
Telemetry is a Solana trading and analytics terminal by BONKbot, offering web and mobile trading, token discovery, wallet tracking, and referral/cashback rewards.

##### methodology (what is being counted as tvl, how is tvl being calculated):
N/A; this PR adds a fees adapter, not a TVL adapter.

##### Does this project have a referral program?
Yes

## Changes
- Added `fees/telemetry.ts` as a new v2 Dune-backed fees adapter.
- Added Solana chain config with Telemetry's main fee wallet, rewards account, and public launch start date of `2025-08-12`.
- Reports SOL-denominated `dailyFees`, `dailyUserFees`, `dailyRevenue`, `dailyProtocolRevenue`, and `dailySupplySideRevenue`.
- Added breakdown labels and methodology for trading fees, retained protocol revenue, and referral/cashback rewards.

## Methodology
- Fees are counted as native SOL inflows to Telemetry's main fee wallet.
- Bot-trade protection excludes transactions where the fee wallet itself appears as the trader in `dex_solana.trades`.
- Supply-side revenue counts SOL referral/cashback rewards paid from the fee wallet to the Telemetry rewards account.
- Revenue and protocol revenue are calculated as fees minus referral/cashback rewards.
- Uses Dune `solana.account_activity` for native SOL balance changes, with timestamp/date filters for the requested fetch window.

## Data Quality / Risk
- Telemetry is related to BONKbot, so overlap with the existing `fees/bonk-bot` adapter was checked. The BONKbot adapter uses Dune's curated `bonkbot_solana.bot_trades` table and does not include fee receiver addresses.
- A sample overlap check for `2026-05-10` found no transaction overlap between BONKbot rows and Telemetry fee-wallet inflows.
- The adapter start date uses the public Telemetry launch/rewards window beginning `2025-08-12`.

## Tests
- `pnpm test fees telemetry 2026-05-11`

### CODE REVIEW

File: fees/telemetry.ts [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+  const query = `
+    WITH
+    walletActivity AS (
+      SELECT
+        tx_id,
+        SUM(CASE WHEN address = '${feeWallet}' AND balance_change > 0 THEN balance_change ELSE 0 END) AS fee_lamports,
+        SUM(CASE WHEN address = '${rewardWallet}' AND balance_change > 0 THEN balance_change ELSE 0 END) AS reward_lamports,
+        MAX(CASE WHEN address = '${feeWallet}' THEN 1 ELSE 0 END) AS fee_wallet_rows
+      FROM solana.account_activity
+      WHERE block_date BETWEEN date(from_unixtime(${options.startTimestamp})) AND date(from_unixtime(${options.endTimestamp}))
```
Comment (treeoflife2) [inline fees/telemetry.ts]: duplicate date filters, we have to keep the dune queries as optimal as possible

Comment (treeoflife2) [reply fees/telemetry.ts]: TIME_RANGE filter already adds the date filter

File: fees/telemetry.ts [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+  ProtocolRevenue: {
+    [LABELS.TERMINAL_REVENUE]: "Trading fees retained by Telemetry after referral or cashback rewards.",
+  },
+  SupplySideRevenue: {
+    [LABELS.REWARDS]: "SOL referral or cashback rewards paid from Telemetry's rewards account.",
+  },
+};
+
+const adapter: SimpleAdapter = {
+  version: 2,
```
Comment (treeoflife2) [inline fees/telemetry.ts]: dune adapter has to be always version 1

File: fees/telemetry.ts [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+  fetch,
+  adapter: chainConfig,
+  dependencies: [Dependencies.DUNE],
+  isExpensiveAdapter: true,
+  allowNegativeValue: true,
```
Comment (treeoflife2) [inline fees/telemetry.ts:108]: any reason behind allowing negative values

**Review by treeoflife2 — CHANGES_REQUESTED** [review]

File: fees/telemetry.ts (hunk 3 above)
Comment (Shaileshkhote) [reply fees/telemetry.ts:108]: For the rewards we are calculating the funds send from fee wallet to referral wallet and not actual referral claims, so there might be a edge case where the fee wallet sends more fund in batch to referral wallet and the fees accumulated < fees sent

**Comment by Shaileshkhote** [comment]
@treeoflife2 
Updated, thanks.

Changes made:
- Switched Telemetry to `version: 1` 
- Removed duplicate manual date/month filters and now use `TIME_RANGE` in the Dune query.
- Switched the fee label to `METRIC.TRADING_FEES`.

Regarding the allowNegativeValue
- For the rewards we are calculating the funds send from fee wallet to referral wallet and not actual referral claims, so there might be a edge case where the fee wallet sends more fund in batch to referral wallet and the fees accumulated < fees sent.

Re-tested:

- `pnpm test fees telemetry 2026-05-11`

Result:
- Daily fees: `1.83 k`
- Daily user fees: `1.83 k`
- Daily revenue: `1.10 k`
- Daily protocol revenue: `1.10 k`
- Daily supply side revenue: `735.00`

**Review by treeoflife2 — APPROVED** [review]
